### PR TITLE
Count mutations above all nodes

### LINKED ISSUE #331: Counting genome-wide mutations above a focal node
Doing this helps to look at saltational lineages. By default, it should count the mutational events across all the sites going up to the root, but it could be useful to allow counting up to a specified ancestral node instead.

### PR DESCRIPTION
**Pull request by szhan**
Fix #331

### CODE REVIEW

**Comment by szhan** [comment]
The unit tests look like they are failing for reasons unrelated to the commit.

File: sc2ts/info.py [hunk 1]
```diff
@@ -235,6 +235,44 @@ def get_path_mrca(path1, path2, node_time):
...
+):
+    num_muts = np.zeros(len(tree_nodes_preorder), dtype=np.int32)
+    for node in tree_nodes_preorder:
+        pa = tree_parent_array[node]
+        if pa == -1:
+            num_muts[node] = 0  # Root
+        else:
+            assert num_muts[pa] >= 0
+            num_muts[node] = num_muts[pa]
+            num_muts[node] += sum(ts_mutations_node == node)
```
Comment (jeromekelleher) [inline sc2ts/info.py:252]: We can do quite a lot better here by passing in

**Comment by jeromekelleher** [comment]
Nice, I think this is the right approach. I'd forgotten about the postorder array, that's a great idea.

We can do quite a bit better here by precomputing the ``nodes_num_mutations`` value like
```
nodes_num_mutations = np.bincount(ts.mutations_node, minlength=ts.num_nodes)
```
and passing this in rather than ``ts.mutations_node``.

File: sc2ts/info.py [hunk 2]
```diff
@@ -235,6 +235,47 @@ def get_path_mrca(path1, path2, node_time):
...
+            num_muts[node] = 0  # Root
+        else:
+            assert num_muts[pa] >= 0
+            num_muts[node] = num_muts[pa]
+            num_muts[node] += nodes_num_mutations[node]
+    return num_muts
+
+
+def get_num_muts(ts):
+    num_muts_all_trees = np.zeros(
```
[the remark this answers is not in the document]
Comment (szhan) [reply sc2ts/info.py:257]: Oh, I did it just to see the breakdown of muts per tree. I do sum them up outside the function. Can add the sum in.

[the remark this answers is not in the document]
Comment (jeromekelleher) [reply sc2ts/info.py:257]: Better to sum as you go. This will be a very big array later on and cause problems

**Comment by szhan** [comment]
Also have to make sure only mutations above a node are counted if they occur within the genomic span of a tree.

**Comment by jeromekelleher** [comment]
That's fine - create a new array per tree, and accumulate that into the running sum

**Comment by szhan** [comment]
I mean

```python
nodes_num_mutations = np.bincount(
        ts.mutations_node,
        minlength=ts.num_nodes,
    )
```

gets the number of mutations across all trees, right? But these number can differ from tree to tree.

**Comment by szhan** [comment]
Anyway, I think it should subset to sites within tree when processing each trees.

**Comment by jeromekelleher** [comment]
Ah yes, crud. Do the bincount on the sites in the tree interval? That should work

**Comment by jeromekelleher** [comment]
So, select the mutations in the interval using a Boolean array, and bincount on the sub array of nodes

**Comment by szhan** [comment]
Yes

**Comment by szhan** [comment]
It looks like `Tree.preorder()` doesn't visit all nodes in a tree all the time.

**Comment by szhan** [comment]
Hmm, there should be no NAMs anywhere in these trees. Not sure why I can get different number of nodes from preorder across trees from a ts simplified from `clipends-v2-mm_4-noexact-f500-clustloc-mrm_2-rw_10-mgs_10-2021-03-04.ts.tsz`?

File: sc2ts/info.py [hunk 3]
```diff
@@ -235,6 +235,48 @@ def get_path_mrca(path1, path2, node_time):
...
+
+
+def get_num_muts(ts):
+    # All nodes msut be present in all trees.
+    num_muts_all_trees = np.zeros(ts.num_nodes, dtype=np.int32)
+    i = 0
+    for tree in ts.trees():
+        tree_nodes_preorder = tree.preorder()
+        assert ts.num_nodes == len(tree_nodes_preorder)
+        tree_site_ids = np.array([s.id for s in tree.sites()], dtype=np.int32)
```
Comment (jeromekelleher) [inline sc2ts/info.py:263]: mutations_position = ts.sites_position[ts.mutations_site]
tree_mutations = (tree.interval.left <= mutations_position) & (mutations_position < tree.interval.right)

bincount(ta.mutations_node[tree_mutations],...)

Comment (jeromekelleher) [reply sc2ts/info.py:263]: Iterating over sites is slow, and np.in is slower than a simple Boolean mask

**Comment by jeromekelleher** [comment]
We don't expect to visit all nodes in all trees, otherwise recombination would have no effect?

**Comment by jeromekelleher** [comment]
I see the point about non ancestral material, but we're still taking different paths through the graph at different genomic locations and I would expect to visit all nodes each time.

I would just remove the assert and move on

**Comment by szhan** [comment]
Sorry, I mean that every node inherits from something at every site across the genome, hence trees.

**Comment by szhan** [comment]
Got it.

**Review by jeromekelleher — APPROVED** [review]
Lgtm

**Comment by szhan** [comment]
Can also get rid of this:

```python
        if pa == -1:
            num_muts[node] = 0  # Root
        else:
            assert num_muts[pa] >= 0
```

**Comment by szhan** [comment]
I think the tests should cover most cases. I'm not covering edge cases like a line tree or mutations above the root.

**Comment by szhan** [comment]
It doesn't count any mutation above the root. I think it is okay?

**Comment by szhan** [comment]
Modified `_get_num_muts` to get mutations above node 0.
